Status classes

colinodell opened the pull request:
**Please note that this includes the same commit made in #8, as it is a dependency of this functionality.**  I'd recommend reviewing and merging that first.

---

Per [RFC 7231 Section 6](https://tools.ietf.org/html/rfc7231#section-6):

> a client MUST understand the class of any status code, as indicated by the
> first digit, and treat an unrecognized status code as being equivalent to
> the x00 status code of that class

This commit provides the functionality and constants needed to determine and
work with those classes.

colinodell committed:
Add getResponseClass() method and related constants

Per [RFC 7231 Section 6]:

 > a client MUST understand the class of any status code, as indicated by the
 > first digit, and treat an unrecognized status code as being equivalent to
 > the x00 status code of that class

This commit provides the functionality and constants needed to determine and
work with those classes.

[RFC 7231 Section 6]: https://tools.ietf.org/html/rfc7231#section-6

colinodell commented:
This is also an alternative approach to #6.  The advantage of this implementation is that you have a single method to determine the class of an unknown code - there's no need to call 5 different methods to determine the class, making it slightly easier to work with in some cases.

lukasoppermann commented:
Hey @colinodell, I really like the simplicity of this approach, I will review it later. Thanks.

nyamsprod commented:
@colinodell  I'm okay with this approach as well if @lukasoppermann prefers it for simplicity I'll retract/close my previous PR

lukasoppermann commented:
Hey @nyamsprod I think I prefer The simplicity of this approach. Thanks for your contribution none the less, I really appreciate it.

lukasoppermann commented:
Hey @colinodell don't you think we should rather return the class (e.g. `Informational`) instead if the first digit? Since I basically already have the first digit when I provide the 3 digits to the method?

colinodell commented:
The digits correspond directly to the class constants, so you can do things like this:

``` php
$code = $someResponse->getStatusCode();
if ($httpstatus->getResponseClass($code) === Httpstatus::CLASS_SERVER_ERROR) {
    $this->log('Remote server is down!');
}
```

We could return the class constants instead, like this:

``` php
public function getResponseClass($statusCode)
{
    $statusCode = $this->filterHttpStatusCode($statusCode);
    $firstDigit = (int) substr($statusCode, 0, 1);

    switch ($firstDigit) {
        case 1: return self::CLASS_INFORMATIONAL;
        case 2: return self::CLASS_SUCCESS;
        case 3: return self::CLASS_REDIRECTION;
        case 4: return self::CLASS_CLIENT_ERROR;
        case 5: return self::CLASS_SERVER_ERROR;
    }
}
```

The actual returned values will be **exactly** the same as they are now; this is just more verbose and makes it super obvious that you can use the `CLASS_` constants to check the results.

If you'd prefer this implementation I'd be happy to make the change - just let me know.  Otherwise, if you'd rather go with @nyamsprod's implementation that's okay too :)

lukasoppermann commented:
Hey @colinodell, I really like the idea with the cases, as it makes it extremely obvious, as you said. I would appreciate the change and will merge it afterwards. Thanks.
Will you be adding this method to the readme-docs or would you prefer me doing it?

nyamsprod commented:
@lukasoppermann the use of a switch in this perspective seems to be a duplicate... since the computed `$firstDigit` already returns the value of the constants @colinodell  added to the class. What is the added value of the switch ? PHPUnit's `assertSame` won't be able to make the difference between the two versions.

lukasoppermann commented:
Well, the idea is, that the person looking at the code has the ability to understand better what happens and understands that the following works fine:

```
$code = $someResponse->getStatusCode();
if ($httpstatus->getResponseClass($code) === Httpstatus::CLASS_SERVER_ERROR) {
    $this->log('Remote server is down!');
}
```

@colinodell we need to document this in the readme anyway, so I would say we do it the way you would prefer, I am fine with the switch-case but also happy with the shorter version. So please let me know if you would like to change the PR or leave it as is.

nyamsprod commented:
In that case, the short version is fine as it is IMHO. Bottom line, the issue here is not the code but its documentation :+1:

colinodell commented:
The more I think about it, the more I like the `switch`-based approach, so I'll switch to using that.

Here's a related question - should we change the values for those `CLASS_` constants to something else?  Perhaps string values like `1xx` / `2xx` / etc?  Or do you think the current values (which happen to be the first digit) are okay?

colinodell commented:
> so I'll **switch** to using that.

I swear that was totally unintentional :P

lukasoppermann commented:
@colinodell the way I see it we have 3 options
- first digit e.g. `1`
- a text string e.g. `1xx`
- a text string like `Informational`

Personally I think the third option is the best as you could still do the check with the CONSTANTS, but you could also get a meaningful word (words) if desired. Do you see any downside to this?

If the user wants the digit one can easily do it by hand `(int) substr($statusCode, 0, 1);`.

nyamsprod commented:
@colinodell @lukasoppermann I may be wrong about it but in PHP land (PDO, SPL, etc...) constants tend to be integer. A constant should not have a meaning outside of its own usage. You should not rely on its value outside of the class usage. so `Httpstatus::CLASS_INFORMATIONAL` could be equals to 300. That's ok as long as it is understandable by the ``Httpstatus` adding more informations to the constant would render outside code dependent on its value. So I'd go for `1`.

colinodell commented:
Great point @nyamsprod, `1` it is :)  I'll go ahead and update the code now.

colinodell commented:
@lukasoppermann @nyamsprod: I've updated the code to use `switch` and also added a short example to the `README`.  Let me know if you'd like to see any other changes!

nyamsprod commented:
@colinodell this is me nitpicking http://mtdowling.com/blog/2014/03/17/hash-lookups-over-array-search/ TL;DR ..

``` php
public function getResponseClass($statusCode)
{
    $statusCode = $this->filterHttpStatusCode($statusCode);
    $list = [
        1 => self::CLASS_INFORMATIONAL,
        2 => self::CLASS_SUCCESS,
        3 => self::CLASS_REDIRECTION,
        4 => self::CLASS_CLIENT_ERROR,
        5 => self::CLASS_SERVER_ERROR,
    ];
    return $list[(int) substr($statusCode, 0, 1)];
}
```

lukasoppermann commented:
@colinodell thanks for the PR & update, its fine like this. @nyamsprod thanks for your helpful input.
